File: freighter/py/freighter/encoder.py
#  Copyright 2023 Synnax Labs, Inc.
#
#  Use of this software is governed by the Business Source License included in the file
#  licenses/BSL.txt.
#
#  As of the Change Date specified in that file, in accordance with the Business Source
#  License, use of this software will be governed by the Apache License, Version 2.0,
#  included in the file licenses/APL.txt.

from typing import Protocol, Type

import msgpack
from alamos import Instrumentation, trace

from freighter.transport import P, Payload


class EncoderDecoder(Protocol):
    """Protocol for an entity that encodes and decodes values from binary."""

    def content_type(self) -> str:
        """:returns: the HTTP content type of the encoder"""
        ...

    def encode(self, data: Payload) -> bytes:
        """Encodes the given data into a binary representation.
        :param data: The data to encode.
        :returns: The binary representation of the data.
        """
        ...

    def decode(self, data: bytes, pld_t: Type[P]) -> P:
        """Decodes the given binary into a type checked payload.
        :param data: THe binary to decode.
        :param pld_t: The type of the payload to decode into.
        """
        ...


class MsgpackEncoder(EncoderDecoder):
    """A Msgpack implementation of EncoderDecoder."""

    def content_type(self):
        return "application/msgpack"

    def encode(self, payload: Payload) -> bytes:
        return msgpack.packb(payload.dict())  # type: ignore

    def decode(self, data: bytes, pld_t: Type[P]) -> P:
        return pld_t.parse_obj(msgpack.unpackb(data))


class JSONEncoder(EncoderDecoder):
    """A JSON implementation of EncoderDecoder."""

    STRING_ENCODING = "utf-8"

    def content_type(self):
        return "application/json"

    def encode(self, payload: Payload) -> bytes:
        return payload.json().encode()

    def decode(self, data: bytes, pld_t: Type[P]) -> P:
        return pld_t.parse_raw(data.decode(JSONEncoder.STRING_ENCODING))


ENCODER_DECODERS: list[EncoderDecoder] = [
    JSONEncoder(),
    MsgpackEncoder(),
]


class TracingEncoderDecoder(EncoderDecoder):
    """Injects tracing information into the context of a request."""

    wrapped: EncoderDecoder
    instrumentation: Instrumentation

    def __init__(self, wrapped: EncoderDecoder):
        self.wrapped = wrapped

    def content_type(self):
        return self.wrapped.content_type()

    @trace("debug")
    def encode(self, payload: Payload) -> bytes:
        return self.wrapped.encode(payload)

    @trace("debug")
    def decode(self, data: bytes, pld_t: Type[P]) -> P:
        return self.wrapped.decode(data, pld_t)

File: x/ts/src/binary/encoder.ts
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { type z, type ZodSchema } from "zod";

import { caseconv } from "@/caseconv";
import { isObject } from "@/identity";

/**
 * Codec is an entity that encodes and decodes messages to and from a
 * binary format.
 */
export interface Codec {
  /** The HTTP content type of the encoder */
  contentType: string;

  /**
   * Encodes the given payload into a binary representation.
   *
   * @param payload - The payload to encode.
   * @returns An ArrayBuffer containing the encoded payload.
   */
  encode: (payload: unknown) => ArrayBuffer;

  /**
   * Decodes the given binary representation into a type checked payload.
   *
   * @param data - The data to decode.
   * @param schema - The schema to decode the data with.
   */
  decode: <P>(data: Uint8Array | ArrayBuffer, schema?: ZodSchema<P>) => P;
}

/** JSONCodec is a JSON implementation of Codec. */
export class JSONCodec implements Codec {
  contentType = "application/json";
  private readonly decoder: TextDecoder;
  private readonly encoder: TextEncoder;

  constructor() {
    this.decoder = new TextDecoder();
    this.encoder = new TextEncoder();
  }

  encode(payload: unknown): ArrayBuffer {
    return this.encoder.encode(this.encodeString(payload)).buffer as ArrayBuffer;
  }

  decode<P extends z.ZodTypeAny>(
    data: Uint8Array | ArrayBuffer,
    schema?: P,
  ): z.output<P> {
    return this.decodeString(this.decoder.decode(data), schema);
  }

  decodeString<P extends z.ZodTypeAny>(data: string, schema?: P): z.output<P> {
    const parsed = JSON.parse(data);
    const unpacked = caseconv.snakeToCamel(parsed);
    return schema != null ? schema.parse(unpacked) : (unpacked as z.output<P>);
  }

  encodeString(payload: unknown): string {
    const caseConverted = caseconv.camelToSnake(payload);
    return JSON.stringify(caseConverted, (_, v) => {
      if (ArrayBuffer.isView(v)) return Array.from(v as Uint8Array);
      if (isObject(v) && "encode_value" in v) {
        if (typeof v.value === "bigint") return v.value.toString();
        return v.value;
      }
      if (typeof v === "bigint") return v.toString();
      return v;
    });
  }

  static registerCustomType(): void {}
}

/**
 * CSVCodec is a CSV implementation of Codec.
 */
export class CSVCodec implements Codec {
  contentType = "text/csv";

  encode(payload: unknown): ArrayBuffer {
    const csvString = this.encodeString(payload);
    return new TextEncoder().encode(csvString).buffer as ArrayBuffer;
  }

  decode<P extends z.ZodTypeAny>(
    data: Uint8Array | ArrayBuffer,
    schema?: P,
  ): z.output<P> {
    const csvString = new TextDecoder().decode(data);
    return this.decodeString(csvString, schema);
  }

  encodeString(payload: unknown): string {
    if (!Array.isArray(payload) || payload.length === 0 || !isObject(payload[0]))
      throw new Error("Payload must be an array of objects");

    const keys = Object.keys(payload[0]);
    const csvRows = [keys.join(",")];

    payload.forEach((item: any) => {
      const values = keys.map((key) => JSON.stringify(item[key] ?? ""));
      csvRows.push(values.join(","));
    });

    return csvRows.join("\n");
  }

  decodeString<P extends z.ZodTypeAny>(data: string, schema?: P): z.output<P> {
    const [headerLine, ...lines] = data
      .trim()
      .split("\n")
      .map((line) => line.trim());
    if (headerLine.length === 0)
      return schema != null ? schema.parse({}) : ({} as z.output<P>);
    const headers = headerLine.split(",").map((header) => header.trim());
    const result: { [key: string]: any[] } = {};

    headers.forEach((header) => {
      result[header] = [];
    });

    lines.forEach((line) => {
      const values = line.split(",").map((value) => value.trim());
      headers.forEach((header, index) => {
        const v = this.parseValue(values[index]);
        if (v == null) return;
        result[header].push(v);
      });
    });

    return schema != null ? schema.parse(result) : (result as z.output<P>);
  }

  private parseValue(value?: string): any {
    if (value == null || value.length === 0) return null;
    const num = Number(value);
    if (!isNaN(num)) return num;
    if (value.startsWith('"') && value.endsWith('"')) return value.slice(1, -1);
    return value;
  }

  static registerCustomType(): void {}
}

export class TextCodec implements Codec {
  contentType = "text/plain";

  encode(payload: unknown): ArrayBuffer {
    return new TextEncoder().encode(payload as string).buffer as ArrayBuffer;
  }

  decode<P extends z.ZodTypeAny>(
    data: Uint8Array | ArrayBuffer,
    schema?: P,
  ): z.output<P> {
    const text = new TextDecoder().decode(data);
    return schema != null ? schema.parse(text) : (text as z.output<P>);
  }
}

export const JSON_CODEC = new JSONCodec();
export const CSV_CODEC = new CSVCodec();
export const TEXT_CODEC = new TextCodec();

export const ENCODERS: Codec[] = [JSON_CODEC];
